sprint-3/step-1

ha308ing opened the pull request:
# Спринт 3. Шаг 1.

## Задачи

- настроить роутинг и авторизацию
  - авторизованный пользователь не может открыть: логин, регистрацию, восстановление пароля
  - неавторизованный пользоатель не может открыть страницу профиля и её дочерние страницы
  - страница с восстановлением пароля доступна только при переходе со страницы с запросом почты для кода
  - при клике на ингридиент открывается модальное окно и меняется адрес, при переходе на этот адрес напрямую открывается отдельная страница с ингридиентом
  - если сохранён актуальный токен доступа, восстанавливать доступ без авторизации

## Демо

[![2024-06-18-00-07-22.gif](https://i.postimg.cc/65h1DTKj/2024-06-18-00-07-22.gif)](https://postimg.cc/svvcWjv7)

[Ссылка на гифку](https://postimg.cc/svvcWjv7)

- **sprint-2/step-1 (#3)**
- **chore(deps): add react-router-dom**
- **feat(router): router infra and home page**
- **refactor(layout): add LayoutMain**
- **feat(layout-form): add form layout and components**
- **feat(pages): add pages and routes**
- **feat(api): update api**
- **feat(store): add forms slices**
- **feat(store): add profile slice**
- **refactor(store): ingredients, order slices**
- **refactor(components): layouts**
- **refactor(components): modals**
- **refactor(components): app-header**
- **refactor(components): button-checkout**
- **refactor(components): ingredients**
- **refactor(component): order-details**
- **feat(pages): profile edit page**
- **feat(pages): ingredient**
- **feat(pages): not-found**
- **feat(pages): login, register**
- **feat(pages): passwords**
- **feat(pages): home**
- **chore(pages): mocks and export**
- **feat(routing): protected-route**
- **feat(routing): update app**
- **refactor(store): update root reducer**
- **chore(styles): add variables**
- **chore(types): add ingredient**
- **chore(components): export**

ha308ing committed:
chore(deps): add react-router-dom

ha308ing committed:
feat(router): router infra and home page

- move app to HomePage
- put routes to app
- add BrowserRouter to index
- routes constants

ha308ing committed:
refactor(layout): add LayoutMain

- move heading from header to home
- fix styles

ha308ing committed:
feat(layout-form): add form layout and components

ha308ing committed:
feat(pages): add pages and routes

- register
- login
- forgot-password
- reset-password
- add routes to app

ha308ing committed:
feat(api): update api

- add new methods
- add custom error and strings translator
- update routes

ha308ing committed:
feat(store): add forms slices

- use form slices for edits and submits

ha308ing committed:
feat(store): add profile slice

ha308ing committed:
refactor(store): ingredients, order slices

ha308ing committed:
refactor(components): layouts

ha308ing committed:
refactor(components): modals

ha308ing committed:
refactor(components): app-header

- use links and store

ha308ing committed:
refactor(components): button-checkout

- use store and routing

ha308ing committed:
refactor(components): ingredients

- use links and routing

ha308ing committed:
refactor(component): order-details

- use states

ha308ing committed:
feat(pages): profile edit page

ha308ing committed:
feat(pages): ingredient

ha308ing committed:
feat(pages): not-found

ha308ing committed:
feat(pages): login, register

ha308ing committed:
feat(pages): passwords

ha308ing committed:
feat(pages): home

ha308ing committed:
chore(pages): mocks and export

ha308ing committed:
feat(routing): protected-route

ha308ing committed:
feat(routing): update app

ha308ing committed:
refactor(store): update root reducer

- add slices
- add global reset

ha308ing committed:
chore(styles): add variables

ha308ing committed:
chore(types): add ingredient

ha308ing committed:
chore(components): export

ha308ing committed:
sprint-2/step-1 (#3)

- add redux store
- get ingredients and send order with api
- drag and drop
- ingredients scroll indicator

ha308ing committed:
fix: update with master

ha308ing committed:
fix: netlify redirects

ha308ing committed:
fix: ingredient details get from params

ha308ing committed:
refactor(ingredient): drag->link->ingredient

ha308ing committed:
chore: readme s3s1

gennady-bars commented on src/services/profile/thunks/logout.js:
```diff
@@ -0,0 +1,7 @@
+import { createAsyncThunk } from "@reduxjs/toolkit";
+import { burgersApiController } from "../../../utils/api/burgers-api-controller";
+
+export const logout = createAsyncThunk("profile/logout", async () => {
+  const logoutStatus = await burgersApiController.logout();
+  return logoutStatus;
+});
```
**Можно лучше**

* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде


```js

export const logout = createAsyncThunk("profile/logout", burgersApiController.logout);

```

gennady-bars commented on src/services/ingredients/thunks/get-ingredients-thunk.js:
```diff
@@ -1,7 +1,7 @@
+import { burgersApiController } from "../../../utils/api/burgers-api-controller";
 
 export const getIngredientsThunk = createAsyncThunk(
   "ingredients/get",
-  burgersApiService.fetchIngredients,
+  async () => burgersApiController.getIngredients,
```
` async () => burgersApiController.getIngredients,`

Тут нет вызова

Надо было оставить то, что было в прошлый раз


```js

export const getIngredientsThunk = createAsyncThunk(
  "ingredients/get",
  burgersApiController.getIngredients
);
```


Вы просто сделали сложную логику в `BurgersApiController`. Не было смысла там вызывать запрос и записывать `requestPromise` в поле `getIngredients`

gennady-bars commented on src/pages/forgot-password/forgot-password.jsx:
```diff
@@ -0,0 +1,70 @@
+      {status === STATUSES.FULFILLED && <ModalFullfilled email={email} />}
+      <LF>
+        <LF.Heading>Восстановление пароля</LF.Heading>
+        <LF.Form onSubmit={handleSubmit}>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/components/protected-route-element/protected-route-element.js:
```diff
@@ -0,0 +1,53 @@
+    return <ModalPending>загрузка</ModalPending>;
+  }
+
+  if (!onlyAuthorized && user) {
+    const from = location.state?.from ?? ROUTES.ROOT;
+    return <Navigate to={from} replace={true} />;
+  }
+
+  if (onlyAuthorized && !user) {
+    return (
+      <Navigate
+        to={ROUTES.LOGIN}
+        state={{ from: location.pathname }}
+        replace={true}
+      />
```
Хорошо сделали `ProtectedRouteElement`

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -1,42 +1,111 @@
+    dispatch(profileActions.get());
   }, [dispatch]);
 
+  const background = location?.state?.background;
```
Отлично сделана логика проверки, кликал ли пользователь сам на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. 
* Хорошо сделали `ProtectedRouteElement`
* Отлично сделана логика проверки, кликал ли пользователь сам на сайте, и показывать ли страницу  с деталями, а не модально окно


**Можно лучше**

* при клике на **логотип** обычно идет переход на **главную** страницу
* Если внутри **безымянной** (стрелочной) функции вызывается одна функция с **точно такими же** аргументами, то эта безымянная функция не нужна. Это лишняя обертка вокруг самой функции, которую можно использовать просто сразу в коде

ha308ing committed:
refactor: api

- remove obsolete abortController
- update service methods
- add postOrder to service
- update thunks

ha308ing committed:
refactor: ingredient details

- remove duplication of useParams
- use clean ingredientDetails in inredient page
- remove withModal since modal forms from url, not ingredient click

ha308ing committed:
chore: rename files, remove obsolete

- remove ingredient slice, active ingredient is not used
- rename initialSlice's, update usage
- remove logs

ha308ing committed:
refactor: app-header

- add app-header layout
- make logo link

ha308ing committed:
refactor: app

- move handlers from app to modal components
- move render condition from reset-password to the app
- update to useRoutes

ha308ing committed:
fix: api service fetch update token check

- update access token if:
  - it's null
  - refresh token is present

ha308ing committed:
refactor: app loading

- add app slice with loading status
- add api service method to load ingredient and profile
- move isAuthChecked from profile slice to app slice
- update usage

ha308ing committed:
fix: fulfilled typo

ha308ing committed:
feat(mobile): add isMobile state to the app

- add media query
- on media query change update the state
- update app slice

ha308ing committed:
feat(mobile): mobile modal component

- add hoc withMobile to render alternative components and mobile modal

ha308ing committed:
refactor(profile): separate menu as component

ha308ing committed:
feat(mobile): mobile logo

ha308ing committed:
feat(mobile): hamburger component

ha308ing committed:
refactor: header links

ha308ing committed:
feat(mobile): use mobile app header

ha308ing committed:
feat(mobile): use mobile ingredients

ha308ing committed:
refactor: main layout

ha308ing committed:
feat(mobile): update global styles

ha308ing committed:
refactor: order summary

ha308ing committed:
fix: not found layout

ha308ing committed:
refactor: modal

ha308ing committed:
refactor: order details

ha308ing committed:
refactor: burger constructor

ha308ing committed:
refactor: ingredient details

ha308ing committed:
feat(mobile): profile edit layout

ha308ing committed:
refactor: withModal hoc

ha308ing committed:
feat(mobile): home

ha308ing committed:
refactor: app header rerender
